state-chain/chains/src/sol/api.rs

```diff
@@ -251,11 +253,36 @@ impl<Environment: SolanaEnvironment> SolanaApi<Environment> {
 		message: Vec<u8>,
 		cf_parameters: Vec<u8>,
 	) -> Result<Self, SolanaTransactionBuildingError> {
+		// Verify the validity of the CCM message before building the call.
+		CcmValidityChecker::is_valid(
+			&CcmChannelMetadata {
+				message: message
+					.clone()
+					.try_into()
+					.expect("This is parsed from bounded vec, therefore the size must fit"),
+				gas_budget: 0, // This value is un-used by Solana
+				cf_parameters: cf_parameters
+					.clone()
+					.try_into()
+					.expect("This is parsed from bounded vec, therefore the size must fit"),
+			},
+			transfer_param.asset.into(),
+		)
+		.map_err(SolanaTransactionBuildingError::InvalidCcm)?;
+
 		let ccm_accounts = SolCcmAccounts::decode(&mut &cf_parameters[..])
 			.map_err(|_| SolanaTransactionBuildingError::CannotDecodeCcmCfParam)?;
```

Comment:
we are decoding the same ```cf_parameters``` two times here. one inside the ```is_valid()`` and one right after it. This is not ideal but I think deduplicating it might take an annoying refactor so I am also ok with leaving it as is.

Reply:
And also we end up using a `ccm_accounts` that is decoded outside the `CcmValidityChecker` so we are not ensuring it's the same. This has very little room for going wrong right now but if we use this in the future for other chains the room for error increases.
We could deduplicate by having the `CcmValidityChecker` return the decoded cf_parameters with a different type per each chain or we could have a Validity check per each chain, but both options require some refactoring.
Another option could be to return an Enum. While it's not great, it adds some safety without much refactoring, something like this:
https://github.com/chainflip-io/chainflip-backend/compare/bc3923fe9362d1dcfc4ab15ba79041ba07307fd4..2f928397dec621e14ac13f75ba72a23f4406d3fe
Or we can just leave it as is and then in the future if/when we use this for other chains we can refactor it as it makes more sense depending on the requirements.

Reply:
Yeah good point. 
I also agree that this is the best approach here. I've made the appropriate changes.

Reply:
looks good now!